### ISSUE TITLE
Verify logical equivalence of large compiled circuits

### ISSUE THREAD

**Issue by jordandsullivan** [unitaryfoundation/ucc#62]
### Background
When we compile a quantum circuit, we need to make sure the resulting circuit is logically equivalent to the input. Our existing [tests to verify logical equivalence](https://github.com/unitaryfoundation/ucc/blob/main/ucc/tests/test_compile.py#L209) of circuits before and after compilation work by comparing the state vectors. As it is exponentially time and memory intensive to simulate the full state vector of a quantum system, this restricts us to running small circuits of <20 qubits if we want our tests to finish promptly. 

There are many options for checking equivalence: graph isomorphism, unwinding SWAP gates, near-Clifford simulation, and more formal methods. 

The most immediately applicable may be from this paper, since it was written to verify the Qiskit compiler specifically by Fred Chong's group: [Giallar: Push-Button Verification for the Qiskit Quantum Compiler](https://arxiv.org/abs/2205.00661). 
@Misty-W found the repo: https://github.com/taorunz/Giallar/tree/master 

Other tools to be aware of are [MQT-QCEC](https://github.com/cda-tum/mqt-qcec), written in C++ and more general.

### Scope and success criteria
In order to complete this issue, you need to demonstrate the use of one of these tools to verify the logical equivalence of a raw quantum circuit of O(100) qubits and O(1000) gates with its UCC compiled equivalent with a runtime <1 hr, _excluding_ compilation time.

#### Steps
1.  Generate a circuit of order ~100 qubits and ~1000 gates using your preferred frontend. We have a handful of functions to generate Qiskit circuits of variable qubit number in this [test script](https://github.com/unitaryfoundation/ucc/blob/main/ucc/tests/test_compile.py), which you may find useful. It may be easiest to prototype using a jupyter notebook or Python script. 
2. Compile your large circuit using UCC: `compiled_circuit = ucc.compile(raw_n100_circuit)`. 
3. Use your chosen tool to compare the logical equivalence of the raw 100 qubit circuit with its compiled version. This step should run in no more than 1hr (and may indeed take far less time than that). 
4. Assuming you are successful, add a test to [ucc/tests/test_compile.py](https://github.com/unitaryfoundation/ucc/blob/main/ucc/tests/test_compile.py) to confirm the logical equivalence of these circuits.

**Comment by ACE07-Sev** [comment]
So, generate a circuit with 100 qubits and 1000 gates, and then time how long it takes to assert its equality to the compiled version?

**Comment by ACE07-Sev** [comment]
What's the time to beat really? I haven't tried Giallar one. Could I know how long that took please if possible?

**Comment by limitless-100** [comment]
@jordandsullivan Is this project still open? If so, I would love to start working on it as soon as possible. Thank you.

**Comment by jordandsullivan** [comment]
Yes @limitless-100 , go ahead!

**Comment by WingCode** [comment]
Hi @jordandsullivan — do you have any tips for accelerating MQT-QCEC’s qcec.verify() equivalence checks? I compiled the library with OpenMPI and ran it on a 96-core machine, yet verifying a 100-qubit random Clifford circuit (from the UCC tests) still takes more than an hour.

**Comment by jordandsullivan** [comment]
Gotcha thanks for the update!

**Comment by WingCode** [comment]
> [@WingCode](https://github.com/WingCode) did you have any luck doing equivalence checking with Giallar?

I ran into scaling issues when trying to extend the [Giallar implementation](https://github.com/WingCode/Giallar/tree/make-importable) to larger qubit counts and estimated that it couldn’t be solved before the hackathon deadline today, so I had to set Giallar aside for now.

**Comment by jordandsullivan** [comment]
@WingCode gotcha. Out of curiosity, did you get a sense of the max account Giallar or MQT-QCEC could handle in under an hour?

**Comment by WingCode** [comment]
MQT-QCEC could handle 20 qubits on random clifford circuits. Greater than 20 qubits wasn't linearly scaling across time in my testing (with the default set of arguments.

Giallar was pretty fast on 20 qubits. But it was running out of memory (64GB) once >20 qubits was being tried.

Pyzx `compare_tensors` runs out of memory on 50 qubits.. Pyzx `verify_equivality` seems promising but it seems to return no equality for the compile and non-compiled version of the circuit

> [@WingCode](https://github.com/WingCode) gotcha cool. Out of curiosity, did you get a sense of the max account Giallar or MQT-QCEC could handle in under an hour?

**Comment by WingCode** [comment]
@bachase / @Misty-W / @bachase 
I stumbled upon [Stim](https://github.com/quantumlib/Stim). According to the paper [here],(https://arxiv.org/abs/2308.01206#:~:text=present%20a%20deterministic%20algorithm%20that,based%20and) it seems to check equivalence of two circuits comprised of Clifford gates by constructing the Tableau (i.e. the final stabilizer mapping of input Pauli generators) and then compare these tableaus directly. This yields an exact yes/no equivalence result in time linear in the number of gates.

**Comment by jordandsullivan** [comment]
@WingCode yes, Stim works for circuits based on only Cliffords, but to my knowledge it doesn't work for general circuits.

**Comment by jordandsullivan** [comment]
@WingCode gotcha. Since

**Comment by jordandsullivan** [comment]
Thanks @WingCode !

Also I just discovered the [Munich Quantum Toolkit](https://mqt.readthedocs.io/en/latest/index.html), which has a whole section of [verification of quantum circuits](https://mqt.readthedocs.io/en/latest/handbook/04_verification.html#verification-of-quantum-circuits). They've have multiple references on doing this efficiently.

**Comment by WingCode** [comment]
> [@WingCode](https://github.com/WingCode) just share that report whenever you get a chance :)

Please find the benchmarking [report](https://colab.research.google.com/drive/15vKRrKPWPNzNpbS0SYfXoW7wCSlqX3fg?usp=sharing) here.
